File: go/internal/bench/fswrite.go
package bench

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"syscall"
	"time"
)

// FSWFocusArea is the focus area for every filesystem-write experiment.
const FSWFocusArea = "filesystem-write"

// SyncKind selects the durability barrier issued per commit.
type SyncKind int

const (
	// SyncFull is a full fsync (data + all metadata).
	SyncFull SyncKind = iota
	// SyncData is fdatasync (data + size, skips timestamps).
	SyncData
)

// FsConfig holds the filesystem-write parameters from the FSW_* env vars.
type FsConfig struct {
	Dir        string
	EntryBytes int
	Warmup     int
	Iterations int
	Batch      int
}

// LoadFsConfig reads FSW_DIR (required), FSW_ENTRY_BYTES, FSW_WARMUP,
// FSW_ITERATIONS and FSW_BATCH, applying defaults. A missing FSW_DIR (guards
// against tmpfs) or any invalid value yields an error.
func LoadFsConfig() (FsConfig, error) {
	dir := strings.TrimSpace(os.Getenv("FSW_DIR"))
	if dir == "" {
		return FsConfig{}, fmt.Errorf("FSW_DIR: required (set FSW_DIR=<dir on a real disk, not tmpfs>)")
	}
	entryBytes, err := positiveEnv("FSW_ENTRY_BYTES", 256)
	if err != nil {
		return FsConfig{}, err
	}
	warmup, err := positiveEnv("FSW_WARMUP", 5000)
	if err != nil {
		return FsConfig{}, err
	}
	iterations, err := positiveEnv("FSW_ITERATIONS", 50000)
	if err != nil {
		return FsConfig{}, err
	}
	batch, err := positiveEnv("FSW_BATCH", 32)
	if err != nil {
		return FsConfig{}, err
	}
	return FsConfig{Dir: dir, EntryBytes: entryBytes, Warmup: warmup, Iterations: iterations, Batch: batch}, nil
}

// RunDurableAppend runs one filesystem-write experiment and returns the per-sync
// latencies in nanoseconds plus the end-to-end throughput in entries/sec.
// batchSize entries are written per sync; prealloc pre-writes the file so a
// size-extending sync is avoided.
func RunDurableAppend(cfg FsConfig, experiment string, sync SyncKind, batchSize int, prealloc bool) ([]int64, float64, error) {
	path := filepath.Join(cfg.Dir, "filesystem-write-"+experiment+".log")
	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0o644)
	if err != nil {
		return nil, 0, fmt.Errorf("open %s: %w", path, err)
	}
	defer f.Close()

	// Make the file's existence durable (file + parent dir), outside timing.
	if err := f.Sync(); err != nil {
		return nil, 0, fmt.Errorf("initial sync: %w", err)
	}
	if err := syncDir(cfg.Dir); err != nil {
		return nil, 0, err
	}

	entry := make([]byte, cfg.EntryBytes)
	for i := range entry {
		entry[i] = 0xAB
	}

	if prealloc {
		if err := preallocate(f, (cfg.Warmup+cfg.Iterations)*cfg.EntryBytes); err != nil {
			return nil, 0, err
		}
	}

	doSync := func() error {
		if sync == SyncFull {
			return f.Sync()
		}
		return syscall.Fdatasync(int(f.Fd()))
	}

	// Warmup (discarded).
	if err := runEntries(f, entry, cfg.Warmup, batchSize, doSync, nil); err != nil {
		return nil, 0, err
	}

	// Measured.
	samples := make([]int64, 0, (cfg.Iterations+batchSize-1)/batchSize)
	tStart := time.Now()
	if err := runEntries(f, entry, cfg.Iterations, batchSize, doSync, &samples); err != nil {
		return nil, 0, err
	}
	throughput := float64(cfg.Iterations) / time.Since(tStart).Seconds()
	return samples, throughput, nil
}

// runEntries writes `entries` entries in chunks of batchSize, syncing once per
// chunk (trailing short chunk allowed). When samples != nil, each sync's elapsed
// ns is appended.
func runEntries(f *os.File, entry []byte, entries, batchSize int, doSync func() error, samples *[]int64) error {
	remaining := entries
	for remaining > 0 {
		count := batchSize
		if remaining < count {
			count = remaining
		}
		for i := 0; i < count; i++ {
			if _, err := f.Write(entry); err != nil {
				return fmt.Errorf("write: %w", err)
			}
		}
		start := time.Now()
		if err := doSync(); err != nil {
			return fmt.Errorf("sync: %w", err)
		}
		if samples != nil {
			*samples = append(*samples, time.Since(start).Nanoseconds())
		}
		remaining -= count
	}
	return nil
}

// preallocate real-zero-writes `total` bytes, fsyncs once, and seeks back to 0
// so the timed loop overwrites already-written blocks (no i_size extension).
func preallocate(f *os.File, total int) error {
	zeros := make([]byte, 1024*1024)
	remaining := total
	for remaining > 0 {
		n := len(zeros)
		if remaining < n {
			n = remaining
		}
		if _, err := f.Write(zeros[:n]); err != nil {
			return fmt.Errorf("preallocate write: %w", err)
		}
		remaining -= n
	}
	if err := f.Sync(); err != nil {
		return fmt.Errorf("preallocate sync: %w", err)
	}
	if _, err := f.Seek(0, io.SeekStart); err != nil {
		return fmt.Errorf("preallocate seek: %w", err)
	}
	return nil
}

// syncDir fsyncs the directory so a newly created file's entry is durable.
func syncDir(dir string) error {
	d, err := os.Open(dir)
	if err != nil {
		return fmt.Errorf("open dir %s: %w", dir, err)
	}
	defer d.Close()
	if err := d.Sync(); err != nil {
		return fmt.Errorf("sync dir %s: %w", dir, err)
	}
	return nil
}

// EmitFS sorts the per-sync samples and emits the four filesystem-write result
// lines. samples is sorted in place.
func EmitFS(experiment string, samples []int64, throughput float64, iterations int) {
	sort.Slice(samples, func(i, j int) bool { return samples[i] < samples[j] })
	nSync := int64(len(samples))
	Emit(Result{FocusArea: FSWFocusArea, Experiment: experiment, Metric: "durable_append_throughput", Value: throughput, Unit: "ops_per_sec", Samples: int64(iterations)})
	Emit(Result{FocusArea: FSWFocusArea, Experiment: experiment, Metric: "sync_p50", Value: float64(Percentile(samples, 50)), Unit: "ns", Samples: nSync})
	Emit(Result{FocusArea: FSWFocusArea, Experiment: experiment, Metric: "sync_p99", Value: float64(Percentile(samples, 99)), Unit: "ns", Samples: nSync})
	Emit(Result{FocusArea: FSWFocusArea, Experiment: experiment, Metric: "sync_mean", Value: Mean(samples), Unit: "ns", Samples: nSync})
}

File: rust/bench-common/src/fswrite.rs
//! filesystem-write durable-append harness, shared by the fsync/fdatasync/
//! prealloc/batch experiments. Each artifact supplies (SyncKind, batch_size,
//! prealloc); the harness owns config, file setup, warmup, the timed loop, and
//! result emission. stdout stays results-only; std-only.

use std::env;
use std::fs::{File, OpenOptions};
use std::io::{self, Seek, SeekFrom, Write};
use std::path::PathBuf;
use std::time::Instant;

use crate::result;
use crate::stats;

/// Focus area for every filesystem-write experiment.
pub const FOCUS_AREA: &str = "filesystem-write";

/// Which durability barrier to issue per commit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SyncKind {
    /// Full fsync (`File::sync_all`) — flushes data + all metadata.
    Full,
    /// fdatasync (`File::sync_data`) — flushes data + size, skips timestamps.
    Data,
}

/// Parsed, validated filesystem-write configuration (`FSW_*`).
#[derive(Debug, Clone)]
pub struct FsConfig {
    /// Directory for the bench file. Required — no default (guards against tmpfs).
    pub dir: PathBuf,
    /// Entry size in bytes.
    pub entry_bytes: usize,
    /// Discarded warmup operations.
    pub warmup: usize,
    /// Measured entries.
    pub iterations: usize,
    /// Entries per group-commit (the `batch` experiment only).
    pub batch: usize,
}

impl FsConfig {
    /// Read configuration from `FSW_*`, applying defaults and validating.
    pub fn from_env() -> Result<FsConfig, String> {
        let dir = match env::var("FSW_DIR") {
            Ok(raw) if !raw.trim().is_empty() => PathBuf::from(raw.trim()),
            _ => {
                return Err(
                    "FSW_DIR: required (set FSW_DIR=<dir on a real disk, not tmpfs>)".to_string(),
                );
            }
        };
        Ok(FsConfig {
            dir,
            entry_bytes: parse_positive("FSW_ENTRY_BYTES", 256)?,
            warmup: parse_positive("FSW_WARMUP", 5_000)?,
            iterations: parse_positive("FSW_ITERATIONS", 50_000)?,
            batch: parse_positive("FSW_BATCH", 32)?,
        })
    }
}

fn parse_positive(name: &str, default: usize) -> Result<usize, String> {
    match env::var(name) {
        Err(_) => Ok(default),
        Ok(raw) => {
            let value: usize = raw.trim().parse().map_err(|_| {
                format!("{name}: invalid value {raw:?} (expected a positive integer)")
            })?;
            if value == 0 {
                return Err(format!("{name}: must be positive, got 0"));
            }
            Ok(value)
        }
    }
}

/// Build config from env, run the experiment, and emit its four result lines.
pub fn run_and_emit(
    experiment: &str,
    sync_kind: SyncKind,
    prealloc: bool,
    batched: bool,
) -> Result<(), String> {
    let cfg = FsConfig::from_env()?;
    let batch_size = if batched { cfg.batch } else { 1 };
    let (samples, throughput) =
        run_durable_append(&cfg, sync_kind, batch_size, prealloc, experiment)
            .map_err(|e| format!("{e}"))?;
    emit_fs(experiment, &samples, throughput, cfg.iterations);
    Ok(())
}

/// Run the durable-append loop. Returns (per-sync latencies in ns, throughput
/// in entries/sec). `batch_size` entries per sync; `prealloc` pre-writes the
/// file so a size-extending sync is avoided.
fn run_durable_append(
    cfg: &FsConfig,
    sync_kind: SyncKind,
    batch_size: usize,
    prealloc: bool,
    experiment: &str,
) -> io::Result<(Vec<u64>, f64)> {
    let path = cfg.dir.join(format!("filesystem-write-{experiment}.log"));
    let mut file = OpenOptions::new()
        .create(true)
        .write(true)
        .truncate(true)
        .open(&path)?;
    // Make the file's existence durable (file + parent dir), outside timing.
    file.sync_all()?;
    if let Some(parent) = path.parent() {
        File::open(parent)?.sync_all()?;
    }

    let entry = vec![0xABu8; cfg.entry_bytes];

    if prealloc {
        let total = (cfg.warmup + cfg.iterations) * cfg.entry_bytes;
        let zeros = vec![0u8; 1024 * 1024];
        let mut remaining = total;
        while remaining > 0 {
            let n = remaining.min(zeros.len());
            file.write_all(&zeros[..n])?;
            remaining -= n;
        }
        file.sync_all()?;
        file.seek(SeekFrom::Start(0))?;
    }

    // Warmup — timings discarded.
    run_entries(&mut file, &entry, cfg.warmup, batch_size, sync_kind, None)?;

    // Measured — record each sync's latency; time the whole span for throughput.
    let mut samples = Vec::with_capacity(cfg.iterations.div_ceil(batch_size));
    let t_start = Instant::now();
    run_entries(
        &mut file,
        &entry,
        cfg.iterations,
        batch_size,
        sync_kind,
        Some(&mut samples),
    )?;
    let throughput = cfg.iterations as f64 / t_start.elapsed().as_secs_f64();

    Ok((samples, throughput))
}

/// Write `entries` entries in chunks of `batch_size`, issuing one sync per chunk
/// (trailing short chunk allowed). When `samples` is `Some`, push each sync's
/// elapsed ns into it.
fn run_entries(
    file: &mut File,
    entry: &[u8],
    entries: usize,
    batch_size: usize,
    sync_kind: SyncKind,
    mut samples: Option<&mut Vec<u64>>,
) -> io::Result<()> {
    let mut remaining = entries;
    while remaining > 0 {
        let count = remaining.min(batch_size);
        for _ in 0..count {
            file.write_all(entry)?;
        }
        let start = Instant::now();
        match sync_kind {
            SyncKind::Full => file.sync_all()?,
            SyncKind::Data => file.sync_data()?,
        }
        if let Some(s) = samples.as_deref_mut() {
            s.push(start.elapsed().as_nanos() as u64);
        }
        remaining -= count;
    }
    Ok(())
}

/// Sort the sync samples and emit the four filesystem-write result lines.
fn emit_fs(experiment: &str, sync_samples: &[u64], throughput: f64, iterations: usize) {
    let mut sorted = sync_samples.to_vec();
    sorted.sort_unstable();
    let n_sync = sorted.len();

    result::emit_float(
        FOCUS_AREA,
        experiment,
        "durable_append_throughput",
        throughput,
        "ops_per_sec",
        iterations,
    );
    result::emit(
        FOCUS_AREA,
        experiment,
        "sync_p50",
        stats::percentile(&sorted, 50.0),
        "ns",
        n_sync,
    );
    result::emit(
        FOCUS_AREA,
        experiment,
        "sync_p99",
        stats::percentile(&sorted, 99.0),
        "ns",
        n_sync,
    );
    result::emit_float(
        FOCUS_AREA,
        experiment,
        "sync_mean",
        stats::mean(sync_samples),
        "ns",
        n_sync,
    );
}

#[cfg(test)]
mod tests {
    use super::*;

    fn scratch_dir(tag: &str) -> std::path::PathBuf {
        let dir = std::env::temp_dir().join(format!("fswrite-test-{}-{tag}", std::process::id()));
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }

    #[test]
    fn batch_chunking_and_prealloc_sync_counts() {
        let dir = scratch_dir("counts");
        let cfg = FsConfig {
            dir: dir.clone(),
            entry_bytes: 64,
            warmup: 5,
            iterations: 20,
            batch: 4,
        };

        // batch 4 over 20 iterations → 5 syncs; throughput positive.
        let (samples, tput) =
            run_durable_append(&cfg, SyncKind::Data, 4, false, "test-batch").unwrap();
        assert_eq!(samples.len(), 5, "20 entries / batch 4 = 5 syncs");
        assert!(tput > 0.0);

        // non-divisible: 21 iterations, batch 4 → ceil = 6 syncs.
        let cfg2 = FsConfig {
            iterations: 21,
            ..cfg.clone()
        };
        let (s2, _) = run_durable_append(&cfg2, SyncKind::Data, 4, false, "test-batch2").unwrap();
        assert_eq!(s2.len(), 6, "ceil(21/4) = 6 syncs");

        // prealloc, batch 1 → 20 syncs; file at least the preallocated size.
        let (s3, _) = run_durable_append(&cfg, SyncKind::Data, 1, true, "test-prealloc").unwrap();
        assert_eq!(s3.len(), 20);
        let len = std::fs::metadata(dir.join("filesystem-write-test-prealloc.log"))
            .unwrap()
            .len();
        assert!(len >= ((cfg.warmup + cfg.iterations) * cfg.entry_bytes) as u64);

        std::fs::remove_dir_all(&dir).ok();
    }
}
